Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+//  NIOHTTP2
+//
+//  Created by Jim Dovey on 11/14/18.
+//
+
+import NIO
+import NIOHPACK
+
+/// A temporary item used to hold onto the HPACK encoder/decoder. Something in the real stream state engine will
+/// take its place soon.
+class HTTP2StreamData {
+    static private var lookup: Dictionary<HTTP2StreamID, HTTP2StreamData> = [:]
+    
+    static func getStream(for streamID: HTTP2StreamID, allocator: ByteBufferAllocator) -> HTTP2StreamData {
+        if let stream = lookup[streamID] {
+            return stream
+        }
+        
+        let stream = HTTP2StreamData(streamID: streamID, allocator: allocator)
+        lookup[streamID] = stream
+        return stream
+    }
+    
+    var encoder: HPACKEncoder
+    var decoder: HPACKDecoder
+    
+    let streamID: HTTP2StreamID
+    
+    init(streamID: HTTP2StreamID, allocator: ByteBufferAllocator) {
+        // TODO: let's pass through a decent value for table sizes here.
+        self.encoder = HPACKEncoder(allocator: allocator)
+        self.decoder = HPACKDecoder(allocator: allocator)
+        self.streamID = streamID
+    }
+    
+    // internal so tests can see it.
+    class func reset() {
+        lookup.removeAll()
+    }
+}
+
+/// Ingests HTTP/2 data and produces frames. You feed data in, and sometimes you'll get a complete frame out.
+class HTTP2FrameDecoder {
+    private var streamDecoders: Dictionary<HTTP2StreamID, HPACKDecoder> = [:]
+    
+    private enum ParserState {
+        case idle
+        case accumulatingData(FrameHeader, ByteBuffer)
+    }
+    
+    private var state: ParserState = .idle
+    private var allocator: ByteBufferAllocator
+    private var currentHeaderFrame: HTTP2Frame? = nil
+    
+    /// Creates a new HTTP2 frame decoder.
+    ///
+    /// - parameter allocator: A `ByteBufferAllocator` used when accumulating blocks of data
+    ///                        and decoding headers.
+    init(allocator: ByteBufferAllocator) {
+        self.allocator = allocator
+    }
+    
+    
+    /// Reads at most one frame's worth of bytes from the input buffer, and returns a decoded
+    /// frame if one has been completely received.
+    ///
+    /// The method's effect on the input buffer and its return value depend on the content of
+    /// the buffer:
+    ///
+    /// - If the input buffer contains more bytes beyond the current frame, those bytes will
+    ///   remain in the buffer; you may continue calling `decode(bytes:)` until the buffer is
+    ///   empty, or until its size does not change.
+    /// - If a given buffer does not contain a complete frame header (9 bytes), this method
+    ///   will take no action, and the buffer will remain unchanged when the method returns.
+    /// - If the buffer does not contain an entire frame including payload, it will be cached
+    ///   internally by the decoder, and the next call to `decode(bytes:)` will append any
+    ///   new bytes to that cache in an effort to complete that frame's payload.
```

Comment:
I think it might be nicer to have some explicit transfer of ownership of the `ByteBuffer`, probably via a different method to this one. My concern is that because we're passing the `ByteBuffer` `inout`, it's not clear that we may be transferring ownership of that buffer to this object, and it's also not immediately clear when exactly that happens.

In fact, my original assumption was that if this method returns `nil` then then `ByteBuffer` is owned by this object, but that's not actually true. Instead, it's if `bytes.readableBytes == 0` at the end of the call.

It might be nicer to have a method to call that unconditionally owns the `ByteBuffer`, and then the user can simply call this method until it returns `nil`. It also makes state management in the connection object a bit easier, as we have a single cumulation buffer when we're single-stepping the state machine.

Reply:
I'm not really transferring the ownership of that buffer, I'm explicitly decoding its bytes into some other types. If I start accumulating bytes, I'm creating my own temporary buffer matching the expected payload size and copying the current bytes into that. I'm specifically not making any assumptions about the byte buffer I'm being given, for example that there's no already-read data in the buffer so I'm not going to keep large amounts of memory around if I just claim ownership.

Additionally, the intention here is to handle the case where the networking stack hands you, the caller, a 4KB buffer containing a HEADERS frame, two CONTINUATIONs, and a DATA frame, all in one buffer. If I'm taking ownership of that buffer in any way, I just took ownership of four frames' worth of data in a function that expects to return at most a single frame.

Chances are that the only frames likely to need to be built out of multiple passed-in buffers are DATA frames, and I'm much happier knowing that I'm only on the hook for the bytes of that frame, not whatever else happened to be allocated for the capacity of some other buffer. I'm reticent to go around purging all previously-read bytes from a given buffer because—since I don't own it—I don't know how it's going to be used beyond here, so I want to avoid the whole 'unnecessarily reallocating a large buffer' issue.

Reply:
> Additionally, the intention here is to handle the case where the networking stack hands you, the caller, a 4KB buffer containing a HEADERS frame, two CONTINUATIONs, and a DATA frame, all in one buffer. If I'm taking ownership of that buffer in any way, I just took ownership of four frames' worth of data in a function that expects to return at most a single frame.

My concern is that *an* entity needs to take ownership of this buffer and manage accumulating within in. If it's not the frame decoder, it has to be the `ConnectionStateMachine`, but one way or another an entity is going to need to own this. This is because the HTTP/2 implementation needs to be a single-step implementation (that is, we parse and operate on one frame at a time) for the sake of correctness.

Given that someone, somewhere needs to take ownership of the buffer my inclination is to have the frame parser do it. I'm willing to be told that you'd prefer that the `ConnectionStateMachine` owns it, but I think that design doesn't really make much sense.